### PR TITLE
MCMC

### PR DESCRIPTION
**Pull request by javdrher**
Adding support for performing Monte-Carlo marginalization of the hyperparameters. This integrates quite nicely into the framework. I haven't tested yet, but it seems that you would be able to construct acquisition hierarchies, mixing point estimates with MCMC. 

There is one minor twist which has to addressed: the following code would currently result in errors:
```python
....
model = GPflow.gpr.GPR(X,Y,RBF(2))
ei = ExpectedImprovement(model)
opt = BayesianOptimizer(domain, ei, hyper_draws=10)
opt.optimize(n_iter=10)
m.likelihood.variance.fixed = True
opt.optimize(n_iter=10)
```
The fix changes the size of the free state, but the copies still maintain the old one. Somehow, I should detect this change, and probably take new copies and replace the operands ParamList. I could verify the length of `get_free_state` but this is not a very robust fix: plenty of situations where no change would be detected. I could verify the position of every Param using the `get_param_index`, but I noticed this method does not take into account the fixed/unfixed state either (is this actually intended GPflow behaviour?) 

Perhaps the easiest way out would be to detect the _needs_recompile flag in _optimize_models and maintain a flag indicating a change in the likelihood graph of the model. This would also catch changes in transforms which would also affect the semantics of the free state. Only changes of prior might introduce some overhead. Any thoughts on this?

### CODE REVIEW

**Comment by javdrher** [comment]
I thought about this, and the rebuilding on a free state change is a broader issue, and is also tightly coupled with the callback for setting model hyperparameters (#7), so I'll build a new PR which builds on this.

File: GPflowOpt/acquisition.py [hunk 1]
```diff
@@ -456,3 +458,42 @@ def __mul__(self, other):
...
+        super(MCMCAcquistion, self).__init__([acquisition] + copies)
+        self._sample_opt = kwargs
+        self._update_hyper_draws()
+
+    def _update_hyper_draws(self):
+        # Sample each model of the acquisition function - results in a list of 2D ndarrays.
+        hypers = np.hstack([model.sample(len(self.operands), **self._sample_opt) for model in self.models])
+
+        # Now visit all copies, and set state
+        for draw, idx in zip(self.operands, range(0, len(self.operands))):
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: for idx, draw in enumerate(self.operands:

?

File: GPflowOpt/acquisition.py [hunk 2]
```diff
@@ -456,3 +458,42 @@ def __mul__(self, other):
...
+            draw.set_state(hypers[idx, :])
+
+    @Acquisition.models.getter
+    def models(self):
+        return self.operands[0].models
+
+    def set_data(self, X, Y):
+        for operand in self.operands:
+            # this triggers model.optimimze() on self.operands[0]
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: model.optimize()

File: GPflowOpt/acquisition.py [hunk 3]
```diff
@@ -456,3 +458,42 @@ def __mul__(self, other):
...
+        assert isinstance(acquisition, Acquisition)
+        assert n_slices > 0
+
+        copies = [copy.deepcopy(acquisition) for _ in range(n_slices - 1)]
+        for c in copies:
+            c._optimize_restarts = 0
+
+        super(MCMCAcquistion, self).__init__([acquisition] + copies)
+        self._sample_opt = kwargs
+        self._update_hyper_draws()
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: is this needed? you need data anyways, and set_data calls the update already

File: GPflowOpt/acquisition.py [hunk 4]
```diff
@@ -456,3 +458,42 @@ def __mul__(self, other):
             return AcquisitionProduct(self.operands.sorted_params + other.operands.sorted_params)
         else:
             return AcquisitionProduct(self.operands.sorted_params + [other])
+
+
+class MCMCAcquistion(AcquisitionSum):
+    def __init__(self, acquisition, n_slices, **kwargs):
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: doc string

I assume acquisition has possibly MLE hyps set?

File: GPflowOpt/acquisition.py [hunk 5]
```diff
@@ -456,3 +458,42 @@ def __mul__(self, other):
...
+        assert isinstance(acquisition, Acquisition)
+        assert n_slices > 0
+
+        copies = [copy.deepcopy(acquisition) for _ in range(n_slices - 1)]
+        for c in copies:
+            c._optimize_restarts = 0
+
+        super(MCMCAcquistion, self).__init__([acquisition] + copies)
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: why is acquisition added? are it's model MLE optimized?

Comment (javdrher) [inline GPflowOpt/acquisition.py]: I could just copy n_slices times, but then some of the inherited tests fail (when calling `models`, the models passed on an AcquisitionAggregation are expected to be returned. So it makes to do it this way.

File: GPflowOpt/acquisition.py (hunk 3 above)
Comment (javdrher) [inline GPflowOpt/acquisition.py]: If not used with `BayesianOptimizer` and models hold data, set_data might not be called. In generally, you'd expect the object to be functional after the constructor.

File: GPflowOpt/acquisition.py (hunk 4 above)
Comment (javdrher) [inline GPflowOpt/acquisition.py]: The call to the parent constructor optimizes the first operand (which has restarts != 0), so it obtains the MLE.

**Pull request merged**